bitcoin/testBwm.c

```diff
@@ -0,0 +1,1063 @@
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+
+#include <inttypes.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "bcash/BRBCashParams.h"
+#include "bitcoin/BRChainParams.h"
+
+#include "BRArray.h"
+#include "BRBIP39Mnemonic.h"
+#include "BRPeerManager.h"
+#include "BRTransaction.h"
+#include "BRWallet.h"
+#include "BRWalletManager.h"
+
+#ifdef __ANDROID__
+#include <android/log.h>
+#define fprintf(...) __android_log_print(ANDROID_LOG_ERROR, "bread", _va_rest(__VA_ARGS__, NULL))
+#define printf(...) __android_log_print(ANDROID_LOG_INFO, "bread", __VA_ARGS__)
+#define _va_first(first, ...) first
+#define _va_rest(first, ...) __VA_ARGS__
+#endif
+
+//
+// Helper Routine
+//
+
+static const char *
+_BRSyncModeString (BRSyncMode m) {
+    switch (m) {
+        case SYNC_MODE_BRD_ONLY:
+        return "SYNC_MODE_BRD_ONLY";
+        case SYNC_MODE_BRD_WITH_P2P_SEND:
+        return "SYNC_MODE_BRD_WITH_P2P_SEND";
+        case SYNC_MODE_P2P_WITH_BRD_SYNC:
+        return "SYNC_MODE_P2P_WITH_BRD_SYNC";
+        case SYNC_MODE_P2P_ONLY:
+        return "SYNC_MODE_P2P_ONLY";
+    }
+}
+
+static const char *
+_BRWalletManagerEventTypeString (BRWalletManagerEventType t) {
+    switch (t) {
+        case BITCOIN_WALLET_MANAGER_CONNECTED:
+        return "BITCOIN_WALLET_MANAGER_CONNECTED";
+
+        case BITCOIN_WALLET_MANAGER_CREATED:
+        return "BITCOIN_WALLET_MANAGER_CREATED";
+
+        case BITCOIN_WALLET_MANAGER_DISCONNECTED:
+        return "BITCOIN_WALLET_MANAGER_DISCONNECTED";
+
+        case BITCOIN_WALLET_MANAGER_SYNC_STARTED:
+        return "BITCOIN_WALLET_MANAGER_SYNC_STARTED";
+
+        case BITCOIN_WALLET_MANAGER_SYNC_PROGRESS:
+        return "BITCOIN_WALLET_MANAGER_SYNC_PROGRESS";
+
+        case BITCOIN_WALLET_MANAGER_SYNC_STOPPED:
+        return "BITCOIN_WALLET_MANAGER_SYNC_STOPPED";
+
+        case BITCOIN_WALLET_MANAGER_BLOCK_HEIGHT_UPDATED:
+        return "BITCOIN_WALLET_MANAGER_BLOCK_HEIGHT_UPDATED";
+    }
+    return "<BITCOIN_WALLET_MANAGER_TYPE_UNKNOWN>";
+}
+
+static const char *
+_BRWalletEventTypeString (BRWalletEventType t) {
+    switch (t) {
+        case BITCOIN_WALLET_CREATED:
+        return "BITCOIN_WALLET_CREATED";
+        case BITCOIN_WALLET_BALANCE_UPDATED:
+        return "BITCOIN_WALLET_BALANCE_UPDATED";
+        case BITCOIN_WALLET_TRANSACTION_SUBMITTED:
+        return "BITCOIN_WALLET_TRANSACTION_SUBMITTED";
+        case BITCOIN_WALLET_FEE_PER_KB_UPDATED:
+        return "BITCOIN_WALLET_FEE_PER_KB_UPDATED";
+        case BITCOIN_WALLET_FEE_ESTIMATED:
+        return "BITCOIN_WALLET_FEE_ESTIMATED";
+        case BITCOIN_WALLET_DELETED:
+        return "BITCOIN_WALLET_DELETED";
+    }
+    return "<BITCOIN_WALLET_TYPE_UNKNOWN>";
+}
```

Comment:
Maybe this in Crypto C as well?  The Swift has this:
```
extension BRWalletManagerEventType: CustomStringConvertible {
    public var description: String {
        switch self {
        case BITCOIN_WALLET_MANAGER_CREATED: return "Created"
        case BITCOIN_WALLET_MANAGER_CONNECTED: return "Connected"
        case BITCOIN_WALLET_MANAGER_DISCONNECTED: return "Disconnected"
        case BITCOIN_WALLET_MANAGER_SYNC_STARTED: return "Sync Started"
        case BITCOIN_WALLET_MANAGER_SYNC_PROGRESS: return "Sync Progress"
        case BITCOIN_WALLET_MANAGER_SYNC_STOPPED: return "Sync Stopped"
        case BITCOIN_WALLET_MANAGER_BLOCK_HEIGHT_UPDATED: return "Block Height Updated"
        default: return "<<unknown>>"
        }
    }
}
```
Ignoring the string differences, there should be duplicate strings. I think in ETH, we are pretty consistent with having `enum -> string` functions.

Reply:
^edit with 'there should not be duplicate string...'

Reply:
I don't follow; are you asking to have the Swift call this new `BRWalletEventTypeString` function once it is in C? I'm not sure why this is Swift at the moment; Swift should have no notion of BTC (or ETH) events.

Reply:
Alas, seems the BTC and ETH 'type -> string' functions are hold overs from when System dealt with BTC and ETH directly, instead of CRYPTO as is now the case.  The System callback logging does do a 'type -> string' for `BRCryptoWalletManagerEventType` and others.

Reply:
I added the BTC ones to the Crypto C. Of note, ETH does not have equivalent C functions.